src/npc.c

```diff
@@ -37,45 +53,480 @@ void npc_iter_no_op(void) {
+                npc->spriteInstanceID = func_802DE0EC(npc->currentAnim.w, animList);
+            } else {
+                npc->spriteInstanceID = func_802DE0EC(npc->currentAnim.w | 0x80000000, animList);
+            }
+        } else {
+            npc->flags |= 2;
+        }
+    }
+
+    npc->shadowIndex = create_shadow_type(0, npc->pos.x, npc->pos.y, npc->pos.z);
+    npc->shadowScale = 1.0f;
+
+    if (gGameStatusPtr->isBattle) {
+        i |= 0x800;
+    }
+    return i;
+}
 
 void npc_create_basic(NpcBlueprint* blueprint) {
     npc_create(blueprint, NULL, FALSE);
 }
 
-void npc_create_standard(NpcBlueprint* blueprint, NpcAnimID animList[]) {
+void npc_create_standard(NpcBlueprint* blueprint, NpcAnimID** animList) {
     npc_create(blueprint, animList, FALSE);
 }
 
 void npc_create_partner(NpcBlueprint* blueprint) {
     npc_create(blueprint, NULL, TRUE);
 }
 
-INCLUDE_ASM(void, "npc", npc_free_by_index, s32 listIndex);
+void npc_free_by_index(s32 listIndex) {
+    Npc* npc;
+    s32 i;
+
+    listIndex &= ~0x800;
+
+    npc = (*gCurrentNpcListPtr)[listIndex];
+    if (npc != NULL) {
+        if (npc->flags) {
+            if (npc->blurBuf != NULL) {
+                heap_free(npc->blurBuf);
+                npc->blurBuf = NULL;
+            }
+
+            if (!(npc->flags & 0x40000000)) {
+                if (!(npc->flags & 0x1000000) && func_802DE5E8(npc->spriteInstanceID)) {
+                    PANIC();
+                }
+            }
+            func_80112328(npc->shadowIndex);
 
-INCLUDE_ASM(void, "npc", npc_free, Npc* npc);
+            for (i = 0; i < 2; i++) {
+                func_8003C428(npc, i);
+            }
+
+            if (npc->flags & 0x100000) {
+                disable_npc_blur(npc);
+            }
+
+            heap_free((*gCurrentNpcListPtr)[listIndex]);
+            (*gCurrentNpcListPtr)[listIndex] = NULL;
+            D_8009A604--;
+        }
+    }
+}
+
+void npc_free(Npc* npc) {
+    s32 i;
+
+    if (npc->blurBuf != NULL) {
+        heap_free(npc->blurBuf);
+        npc->blurBuf = NULL;
+    }
+
+    if (!(npc->flags & 0x40000000)) {
+        if (!(npc->flags & 0x1000000) && func_802DE5E8(npc->spriteInstanceID) != 0) {
+            PANIC();
+        }
+    }
+    func_80112328(npc->shadowIndex);
+
+    for (i = 0; i < 2; i++) {
+        func_8003C428(npc, i);
+    }
+
+    if (npc->flags & 0x100000) {
+        disable_npc_blur(npc);
+    }
+
+    heap_free(npc);
+
+    for (i = 0; i < MAX_NPCS; i++) {
+        if ((*gCurrentNpcListPtr)[i] == npc) {
+            break;
+        }
+    }
+
+    (*gCurrentNpcListPtr)[i] = NULL;
+    D_8009A604--;
+}
 
 Npc* get_npc_by_index(s32 listIndex) {
     return (*gCurrentNpcListPtr)[listIndex & ~0x800];
 }
 
-INCLUDE_ASM(void, "npc", npc_do_world_collision, Npc* npc);
+extern s16 D_8010C97A;
+
+void npc_do_world_collision(Npc* npc) {
+    f32 temp_f0;
+    s32 phi_v0;
+    f32 temp_x;
+    f32 temp_y;
+    f32 temp_z;
+
+    if (npc->flags & 0x40) {
```

Comment:
Not really an official project guidline, but I think this would be easier to read with some empty line separating the conditional blocks at least